### PR TITLE
Dogfood ImmutableFileCollection on production code

### PR DESCRIPTION
**Pull request by acourtneybrown**
Branch build: https://builds.gradle.org/viewQueued.html?itemId=12132979&buildTypeId=Gradle_Check_Stage_BranchBuildAccept_Trigger&tab=queuedBuildOverviewTab

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/file/collections/ImmutableFileCollection.java [hunk 1]
```diff
@@ -72,6 +80,10 @@ public TaskDependency getBuildDependencies() {
             this.files = ImmutableSet.copyOf(files);
         }
 
+        FileOnlyImmutableFileCollection(Iterable<File> files) {
+            this.files = ImmutableSet.copyOf(files);
+        }
+
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/file/collections/ImmutableFileCollection.java]: Nitpick: we could have a single constructor here with `ImmutableSet<File>` as its parameter.

File: subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/NormalizingJavaCompilerTest.groovy [hunk 2]
```diff
@@ -19,8 +19,10 @@ import org.gradle.api.internal.file.collections.ImmutableFileCollection
 import org.gradle.api.tasks.WorkResult
 import org.gradle.api.tasks.compile.CompileOptions
 import org.gradle.util.TestUtil
+import org.gradle.util.UsesNativeServices
 import spock.lang.Specification
 
+@UsesNativeServices
```
Comment (lptr) [inline subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/NormalizingJavaCompilerTest.groovy]: Do we need this?

File: subprojects/language-java/src/test/groovy/org/gradle/api/tasks/compile/CompileOptionsTest.groovy [hunk 3]
```diff
@@ -19,9 +19,11 @@ package org.gradle.api.tasks.compile
 import org.gradle.api.internal.file.FileCollectionInternal
 import org.gradle.api.internal.file.collections.ImmutableFileCollection
 import org.gradle.util.TestUtil
+import org.gradle.util.UsesNativeServices
 import spock.lang.Specification
 import spock.lang.Unroll
 
+@UsesNativeServices
```
Comment (lptr) [inline subprojects/language-java/src/test/groovy/org/gradle/api/tasks/compile/CompileOptionsTest.groovy]: We shouldn't need this.

File: subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/DefaultJvmLanguageCompileSpec.java [hunk 4]
```diff
@@ -63,12 +65,12 @@ public void setTempDir(File tempDir) {
 
     @Override
     public FileCollection getSource() {
-        return source;
+        return ImmutableFileCollection.of(source);
     }
```
Comment (lptr) [inline subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/DefaultJvmLanguageCompileSpec.java]: This method is declared in an internal type, so we should be fine to change its return type to `Set<File>`.

File: subprojects/language-scala/src/test/groovy/org/gradle/api/internal/tasks/scala/NormalizingScalaCompilerTest.groovy [hunk 5]
```diff
@@ -22,9 +22,11 @@ import org.gradle.api.tasks.compile.CompileOptions
 import org.gradle.language.base.internal.compile.Compiler
 import org.gradle.language.scala.tasks.BaseScalaCompileOptions
 import org.gradle.util.TestUtil
+import org.gradle.util.UsesNativeServices
 import spock.lang.Specification
 import spock.lang.Unroll
 
+@UsesNativeServices
```
Comment (lptr) [inline subprojects/language-scala/src/test/groovy/org/gradle/api/internal/tasks/scala/NormalizingScalaCompilerTest.groovy]: Don't do this.

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/internal/resolve/SourceSetNativeDependencyResolver.java [hunk 6]
```diff
@@ -59,21 +59,17 @@ private NativeDependencySet createNativeDependencySet(LanguageSourceSet sourceSe
...
         @Override
         public FileCollection getIncludeRoots() {
-            return empty();
+            return ImmutableFileCollection.of();
         }
 
         @Override
         public FileCollection getLinkFiles() {
-            return empty();
+            return ImmutableFileCollection.of();
```
Comment (lptr) [inline subprojects/platform-native/src/main/java/org/gradle/nativeplatform/internal/resolve/SourceSetNativeDependencyResolver.java]: Should we add an empty-parameter factory method for this guy, I wonder? It would mean one less allocation of an empty `Object[]`. Not a big deal... We could also name it `empty()`, but I'm also buying into keeping it named `of()` for the reasons Guava uses. (Which is that when you add/remove elements, you shouldn't need to change the method name.)

File: subprojects/platform-play/src/main/java/org/gradle/play/internal/run/PlayApplicationRunner.java [hunk 7]
```diff
@@ -70,7 +70,7 @@ private PlayClassloaderMonitorDeploymentDecorator(Deployment delegate, PlayRunSp
         private FileCollection collectApplicationClasspath(PlayRunSpec runSpec) {
             Set<File> applicationClasspath = Sets.newLinkedHashSet(runSpec.getChangingClasspath());
             applicationClasspath.add(runSpec.getApplicationJar());
-            return new SimpleFileCollection(applicationClasspath);
+            return ImmutableFileCollection.of(applicationClasspath);
```
Comment (lptr) [inline subprojects/platform-play/src/main/java/org/gradle/play/internal/run/PlayApplicationRunner.java]: I'm not saying we should spend a ton of time optimizing this, but we could use an `ImmutableSet.builder()` instead of the `LinkedHashSet` here and already pass an `ImmutableSet`.

File: subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayTestPluginTest.groovy [hunk 8]
```diff
@@ -57,6 +58,7 @@ class PlayTestPluginTest extends Specification {
             taskName(_, _) >> { String verb, String object -> "${verb}SomeBinary${object.capitalize()}"}
             taskName(_) >> { String verb -> "${verb}SomeBinary"}
         }
+        _ * binary.jarFile >> jarFile
```
Comment (lptr) [inline subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayTestPluginTest.groovy]: Why is this needed now? Also, let's use explicit numbers of repetitions instead of `_` to avoid leaving unused code in the tests.

File: subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/DefaultJvmLanguageCompileSpec.java [hunk 9]
```diff
@@ -63,12 +65,12 @@ public void setTempDir(File tempDir) {
...
 
     @Override
     public void setSource(FileCollection source) {
-        this.source = source;
+        this.source = source.getFiles();
```
Comment (lptr) [inline subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/DefaultJvmLanguageCompileSpec.java]: Looking at this I think it might be best to avoid even referring to `FileCollection` in this class. We should instead pass a `Set<File>` and make a defensive copy via `ImmutableSet.copyOf()`.

Comment (lptr) [reply subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/DefaultJvmLanguageCompileSpec.java]: It would perhaps be a good idea to standardize on the same collection type: we have `Iterable<File>`, `FileCollection` and `List<File>` used in this class already. I'd say `Iterable<File>` or `Collection<File>` should be fine for the setters. We could use the same for the getters, or be more specific, and return `ImmutableSet`. The goal should be consistency, though.

File: subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayTestPluginTest.groovy (hunk 8 above)
Comment (acourtneybrown) [reply subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayTestPluginTest.groovy]: `SimpleFileCollection` happily took the `null` produced by the `binary` `Mock` from `getJarFile()`.  However, now the `ImmutableFileCollection` throws an NPE, so we actually need to stub out the value.  I've updated it to be the right number for this call.

File: subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/DefaultJvmLanguageCompileSpec.java (hunk 9 above)
Comment (acourtneybrown) [reply subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/DefaultJvmLanguageCompileSpec.java]: Commit b2e3ea6408d swaps in `Iterable<File>` (& `Set<File>`) for `FileCollection`.  There are some ancillary changes necessary because `FileCollection` (& `FileTree`) was used to filter out files.  Let me know if think this change is worth keeping/pursuing.

File: subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayTestPluginTest.groovy (hunk 8 above)
Comment (lptr) [reply subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayTestPluginTest.groovy]: Thanks. Can we also add `0 * _`?

Comment (acourtneybrown) [reply subprojects/platform-play/src/test/groovy/org/gradle/play/plugins/PlayTestPluginTest.groovy]: I added  `0 * _` in 5a90200e4d1, but this whole test class is pretty awful.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/file/collections/ImmutableFileCollectionTest.groovy [hunk 10]
```diff
@@ -32,10 +32,12 @@ import java.util.concurrent.Callable
...
 class ImmutableFileCollectionTest extends Specification {
     def 'can create empty collection'() {
-        ImmutableFileCollection files = ImmutableFileCollection.of()
+        ImmutableFileCollection collection1 = ImmutableFileCollection.of()
+        ImmutableFileCollection collection2 = ImmutableFileCollection.of(new File[0])
 
         expect:
-        files.files.size() == 0
+        collection1.files.size() == 0
+        collection2.files.size() == 0
```
Comment (lptr) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/file/collections/ImmutableFileCollectionTest.groovy]: You could even test that `collection1 is collection2`

File: subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java [hunk 11]
```diff
@@ -143,7 +146,11 @@ public void compile(List<String> files, CompilationUnit cu) {
                         if (shouldProcessAnnotations) {
                             // In order for the Groovy stubs to have annotation processors invoked against them, they must be compiled as source.
                             // Classes compiled as a result of being on the -sourcepath do not have the annotation processor run against them
-                            spec.setSource(spec.getSource().plus(new SimpleFileCollection(stubDir).getAsFileTree()));
+                            Set<File> newSource = ImmutableSet.<File>builder()
+                                .addAll(spec.getSource())
+                                .add(stubDir)
+                                .build();
+                            spec.setSource(ImmutableFileCollection.of(newSource).getAsFileTree().getFiles());
```
Comment (lptr) [inline subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java]: This still looks weird to me. This looks like a sign that `FileOnlyImmutableFileCollection` should override `getAsFileTree()` perhaps? Let's discuss.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/file/collections/ImmutableFileCollectionTest.groovy (hunk 10 above)
Comment (acourtneybrown) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/file/collections/ImmutableFileCollectionTest.groovy]: that's what the next test method does :)

File: subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java (hunk 11 above)
Comment (acourtneybrown) [reply subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java]: after further discussion, we've decided to live with the current code.  as part of future cleanup in the Scala, Java & Groovy compiler classes, the handling of the `source` file collection & classpath should be revisited to remove the roundabout way they are configured.

**Comment by acourtneybrown** [comment]
@lptr PR should be ready for you.  As I mentioned in Slack, I've re-added the `FileCollection source` getters and setters to `JvmLanguageCompileSpec` to avoid breaking the `ThirdPartyPluginsSmokeTest`

**Comment by lptr** [comment]
LGTM.

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=12272857

**Comment by lptr** [comment]
New branch build: https://builds.gradle.org/viewQueued.html?itemId=12324782

**Comment by lptr** [comment]
Branch build after rebasing: https://builds.gradle.org/viewQueued.html?itemId=12344064

**Comment by lptr** [comment]
Build: https://builds.gradle.org/viewQueued.html?itemId=12367635

**Comment by lptr** [comment]
Let's try that again: https://builds.gradle.org/viewQueued.html?itemId=12367919

**Comment by lptr** [comment]
Re-run after unable to delete file on `windows7`: https://builds.gradle.org/viewQueued.html?itemId=12371645

**Comment by lptr** [comment]
Re-run after blood sacrifice on the altar of TeamCity: https://builds.gradle.org/viewQueued.html?itemId=12378965

**Comment by lptr** [comment]
Master accept: https://builds.gradle.org/viewQueued.html?itemId=12411603

**Comment by lptr** [comment]
Build: https://builds.gradle.org/viewQueued.html?itemId=12413756

**Comment by lptr** [comment]
Once again: https://builds.gradle.org/viewQueued.html?itemId=12452748

**Comment by lptr** [comment]
Aaaand once more: https://builds.gradle.org/viewQueued.html?itemId=12456092

**Comment by lptr** [comment]
And again: https://builds.gradle.org/viewQueued.html?itemId=12457820

**Comment by lptr** [comment]
Feedback build, hopefully the last: https://builds.gradle.org/viewQueued.html?itemId=12506265

**Comment by lptr** [comment]
And once more: https://builds.gradle.org/viewQueued.html?itemId=12506407
